hxcopy.1
.TH "HXCOPY" "1" "9 Dec 2008" "7.x" "HTML-XML-utils"
.SH NAME
hxcopy \- copy an HTML file and update its relative links
.SH SYNOPSIS
.B hxcopy
.RB "[\| " \-i
.IR old-URL " \|]"
.RB "[\| " \-o
.IR new-URL " \|]"
.RB "[\| " \-s " \|]"
.RB "[\| " \-v " \|]"
.RI "[\| " file-or-URL " [\| " file-or-URL " \|] \|]"
.SH DESCRIPTION
The
.B hxcopy
command copies its first argument to its second argument, while
updating relative links. The input is assumed to be HTML or XHTML and
may be slightly reformatted in the process.
.PP
If the second argument is omitted,
.B hxcopy
writes to standard output. In this case the option
.B \-o
is required. If the first argument is also omitted,
.B hxcopy
reads from standard input. In this case the option
.B \-i
is required.
.SH OPTIONS
The following options are supported:
.TP 10
.BI \-i " old-URL"
For the purposes of updating relative links, act as if
.I old-URL
is the location from which the input is copied. If this option is
omitted, the actual location of the first argument is used for
calculating relative links.
.TP
.BI \-o " new-URL"
For the purposed of updating relative links, act as if
.I new-URL
is the location to which the input is copied. If this option is
omitted, the actual location of the second argument is used for
calculating relative links.
.TP
.B \-s
Also replace links to self. This effects the treatment of empty URLs,
i.e., links to the document itself that do not explicitly name the
document. Without
.BR \-s ","
implicit links to the document itself (href=""), to a fragment of the
document itself (href="#foo") or to a query over the document itself
(href="?query") are not changed and will thus refer to (a fragment of)
the new document. With
.BR \-s ","
these links are rewritten to refer to (a fragment of) the old document
instead.
.TP
.B \-v
Print the version number and exit immediately.
.SH ENVIRONMENT
To use a proxy to retrieve remote files, set the environment variables
.B http_proxy
and
.BR ftp_proxy "."
E.g.,
.B http_proxy="http://localhost:8080/"
.SH BUGS
Unlike the last argument of
.BR cp (1),
the last argument of
.B hxcopy
must be a file, not a directory.
.PP
The second argument must be a local file. Writing to a URL is not yet
implemented. To work around this, replace
.B "hxcopy file.html http://example.org/file.html"
by
.B "hxcopy -o http://example.org/file.html file.html tmp.html"
and then upload tmp.html to the given URL with some other command,
such as
.BR curl (1).
The first argument, however, may be a URL.
.B hxcopy
will download the given file. (Currently only HTTP is supported.)
.SH EXAMPLE
Assume the HTML file foo.html contains a relative link to
"../bar.html". Here are some examples of commands:
.TP
.B "hxcopy foo.html bar/foo.html"
The file foo.html is copied to ../bar/foo.html and the relative link
to "../bar.html" becomes "../../bar.html".
.TP
.B "hxcopy foo.html ../foo.html"
The file foo.html is copied to ../foo.html and the relative link to
"../bar.html" is rewritten as "bar.html".
.TP
.B "hxcopy -i http://my.org/dir1/foo.html -o http://my.org/foo.html file1.html file2.html"
The file file1.html is copied to file2.html and the relative link to
"../bar.html" is rewritten as "bar.html". A command like this may
be useful to update files that are later uploaded to a server.
.SH "SEE ALSO"
.BR cp (1),
.BR curl (1),
.BR hxwls (1)


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxcopy.c
/* hxcopy -- copy an HTML file and update relative URLs at the same time
 *
 * Copy an HTML file with all URLs that were relative to OLDURL
 * updated to be relative to NEWURL instead. (If the document has a
 * BASE element, only that is updated.) OLDURL and NEWURL may
 * themselves be relative (to the same base URL, which need not be
 * mentioned).
 *
 * Part of HTML-XML-utils, see:
 * http://www.w3.org/Tools/HTML-XML-utils/
 *
 * TO DO: Should it be an option whether URL references of the form
 * "", "#foo" and "?bar" are replaced by "oldurl", "oldurl#foo" and
 * "oldurl?bar"? (See adjust_url().)
 *
 * Created: 5 Dec 2008
 * Author: Bert Bos <bert@w3.org>
 *
 * Copyright © 2008-2012 W3C
 * See http://www.w3.org/Consortium/Legal/2002/copyright-software-20021231
 */

#include "config.h"
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <unistd.h>
#include <errno.h>
#include <stdbool.h>
#if HAVE_STRING_H
#  include <string.h>
#endif
#if HAVE_STRINGS_H
#  include <strings.h>
#endif
#include "export.h"
#include "heap.e"
#include "types.e"
#include "html.e"
#include "scan.e"
#include "url.e"
#include "dict.e"
#include "openurl.e"
#include "errexit.e"

#define same(a, b) ((a) ? ((b) && eq((a), (b))) : !(b))

static bool has_errors = false;		/* Enconutered errors during parsing */
static FILE *out = NULL;		/* Where to write output */
static bool has_base = false;		/* Document has a <BASE> element */
static string newbase;			/* Path from OLDURL to NEWURL */
static bool replace_self = false;	/* Change link to self in link to old */


/* path_from_url_to_url -- compute URL that is path from one URL to another */
static string path_from_url_to_url(const conststring a, const conststring b)
{
  URL p, q;
  string s = NULL;
  char cwd[4096];
  int i, j;

  if (!getcwd(cwd, sizeof(cwd) - 1)) return NULL; /* To do: handle long path */
  strcat(cwd, "/");
  s = URL_s_absolutize(cwd, a); p = URL_new(s); dispose(s);
  s = URL_s_absolutize(cwd, b); q = URL_new(s); dispose(s);
  if (p->proto && !q->proto) {
    errno = EACCES;		/* Path from remote to local not possible */
  } else if (!same(p->proto, q->proto) ||
      !same(p->user, q->user) ||
      !same(p->password, q->password) ||
      !same(p->machine, q->machine) ||
      !same(p->port, q->port)) {
    s = newstring(b);		/* Just use the URL b */
  } else {
    /* Find the last '/' before which both paths are the same */
    for (j = i = 0; p->path[i] && q->path[i] && p->path[i] == q->path[i]; i++)
      if (p->path[i] == '/') j = i;

    /* Construct path from a to b by descending a and climbing b */
    for (i = j + 1; p->path[i]; i++)
      if (p->path[i] == '/') strapp(&s, "../", NULL);
    strapp(&s, q->path + j + 1, NULL);
  }
  URL_dispose(p);
  URL_dispose(q);
  return s;
}


/* adjust_url -- return a new URL relative to newurl instead of oldurl */
static conststring adjust_url(const conststring url)
{
  if (!replace_self && (!url || !url[0] || url[0] == '#' || url[0] == '?'))
    return url;			/* Don't replace references to self */
  else
    return URL_s_absolutize(newbase, url);
}


/* attribute_is_url -- check if the attribute is URL-valued */
static bool attribute_is_url(const conststring attrib)
{
  return strcasecmp(attrib, "href") == 0 ||
    strcasecmp(attrib, "src") == 0 ||
    strcasecmp(attrib, "action") == 0 ||
    strcasecmp(attrib, "background") == 0 ||
    strcasecmp(attrib, "cite") == 0 ||
    strcasecmp(attrib, "classid") == 0 ||
    strcasecmp(attrib, "codebase") == 0 ||
    strcasecmp(attrib, "data") == 0 ||
    strcasecmp(attrib, "longdesc") == 0 ||
    strcasecmp(attrib, "profile") == 0 ||
    strcasecmp(attrib, "usemap") == 0;
}


/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
  has_errors = true;
}


/* start -- called before the first event is reported */
void* start(void)
{
  return NULL;
}

  
/* end -- called after the last event is reported */
void end(void *clientdata)
{
  /* skip */
}


/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, string commenttext)
{
  fprintf(out, "<!--%s-->", commenttext);
}


/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, string text)
{
  fprintf(out, "%s", text);
}


/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, string gi,
		 string fpi, string url)
{
  fprintf(out, "<!DOCTYPE %s", gi);
  if (fpi) fprintf(out, " PUBLIC \"%s\"", fpi);
  if (url) fprintf(out, " %s\"%s\"", fpi ? "" : "SYSTEM ", url);
  fprintf(out, ">");
}


/* handle_pi -- called after a PI is parsed */
void handle_pi(void *clientdata, string pi_text)
{
  fprintf(out, "<?%s>", pi_text);
}


/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, string name, pairlist attribs)
{
  conststring v;
  pairlist p;

  fprintf(out, "<%s", name);
  for (p = attribs; p; p = p->next) {
    fprintf(out, " %s", p->name);
    if (!p->value) v = NULL;
    else if (has_base) v = newstring(p->value);	/* No need to adjust */
    else if (attribute_is_url(p->name)) v = adjust_url(p->value);
    else v = newstring(p->value);		/* No need to adjust */
    if (v) fprintf(out, "=\"%s\"", v);
    dispose(v);
  }
  fprintf(out, ">");

  /* If this is a <BASE> tag, no further adjustments are needed */
  if (strcasecmp(name, "base") == 0) has_base = true;
}


/* handle_emptytag -- called after an empty tag is parsed */
void handle_emptytag(void *clientdata, string name, pairlist attribs)
{
  conststring v;
  pairlist p;

  fprintf(out, "<%s", name);
  for (p = attribs; p; p = p->next) {
    fprintf(out, " %s", p->name);
    if (!p->value) v = NULL;
    else if (has_base) v = newstring(p->value);	/* No need to adjust */
    else if (attribute_is_url(p->name)) v = adjust_url(p->value);
    else v = newstring(p->value);		/* No need to adjust */
    if (v) fprintf(out, "=\"%s\"", v);
    dispose(v);
  }
  fprintf(out, " />");

  /* If this is a <BASE> tag, no further adjustments are needed */
  if (strcasecmp(name, "base") == 0) has_base = true;
}


/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, string name)
{
  fprintf(out, "</%s>", name);
}


/* usage -- print usage message and exit */
static void usage(const conststring progname)
{
  fprintf(stderr, "Usage: %s [-v] [-s] [-i old-URL] [-o new-URL] [URL [URL]]\n", progname);
  exit(1);
}


int main(int argc, char *argv[])
{
  int c, status = 200;
  string oldurl = NULL, newurl = NULL;

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(start);
  set_end_handler(end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  /* Parse command line */
  while ((c = getopt(argc, argv, "i:o:sv")) != -1)
    switch (c) {
    case 'o': newurl = optarg; break;
    case 'i': oldurl = optarg; break;
    case 's': replace_self = true; break;
    case 'v': printf("Version: %s %s\n", PACKAGE, VERSION); return 0;
    default: usage(argv[0]);
    }
  if (argc > optind + 2) usage(argv[0]);
  if (argc > optind + 1) out =  fopenurl(argv[optind+1], "w", NULL);
  else if (newurl) out = stdout;
  else errexit("%s: option -o is required if output is to stdout\n", argv[0]);
  if (!out) {perror(argv[optind+1]); exit(3);}
  if (argc > optind) yyin = fopenurl(argv[optind], "r", &status);
  else if (oldurl) yyin = stdin;
  else errexit("%s: option -i is required if input is from stdin\n", argv[0]);
  if (!yyin) {perror(argv[optind]); exit(2);}
  if (status != 200) errexit("%s : %s\n", argv[1], http_strerror(status));
  if (!oldurl) oldurl = argv[optind];
  if (!newurl) newurl = argv[optind+1];
  newbase = path_from_url_to_url(newurl, oldurl);
  if (!newbase) errexit("%s: could not parse argument as a URL\n", argv[0]);
  if (yyparse() != 0) exit(4);
  return has_errors ? 1 : 0;
}
